### PR TITLE
Use old _execute_runtime_service for observables that are not compatible with SPO representation

### PR DESCRIPTION
**Pull request by austingmhuang**
Although most observables supported by PL do have a pauli rep (therefore an SPO rep), there are certain observables such as Hadamard that do not have a pauli_rep. This is a problem because we can't translate it to an SPO rep to be used with the Estimator primitive. This PR seeks to amend this by having those circuits sent to the old _execute_runtime_service instead, where we can avoid this problem.

### CODE REVIEW

**Comment by austingmhuang** [comment]
[sc-62046]

**Review by lillian542 — COMMENTED** [review]
Thanks @austingmhuang! I think this looks good, but I think it might be preferable to have the logic checking for a `pauli_rep` with the rest of the sorting logic in the transform that divides things into tapes for `Sampler`, `Estimator` and the legacy runtime. That's in the same fine, in the `split_measurement_types` transform. 

Then we can avoid sending Estimator-compatible measurements to the legacy runtime just because they share a tape with incompatible measurements, and we can avoid having sorting logic distributed different places in the code, which will make debugging harder in the future.

File: pennylane_qiskit/qiskit_device2.py [hunk 1]
```diff
@@ -495,7 +495,9 @@ def execute_circuits(session):
                             f"Setting shot vector {circ.shots.shot_vector} is not supported for {self.name}."
                             f"The circuit will be run once with {circ.shots.total_shots} shots instead."
                         )
-                    if isinstance(circ.measurements[0], (ExpectationMP, VarianceMP)):
+                    if isinstance(circ.measurements[0], (ExpectationMP, VarianceMP)) and all(
+                        mp.obs.pauli_rep for mp in circ.measurements
+                    ):
```
[the remark this answers is not in the document]
Comment (austingmhuang) [reply pennylane_qiskit/qiskit_device2.py:500]: That works as well. Moving the logic into `split_measurement_types` is a good solution that ensures that measurements that can go to the Estimator do go to the estimator. I avoided this solution because (this is just personal preference) it's a bit confusing since the function name suggests that it'll split based on measurement type, but this will split it based on `measurement_type` as well as if the observable has a `pauli_rep`. But I can't think of a different way that allows Estimator-friendly measurements to actually go to the Estimator without splitting it at this place.

I think this is a good solution though, so I will implement it this way and add a comment about this edge case and why it exists.

File: pennylane_qiskit/qiskit_device2.py [hunk 2]
```diff
@@ -514,6 +516,13 @@ def _execute_runtime_service(self, circuits, session):
         # update kwargs in case Options has been modified since last execution
         self._update_kwargs()
 
+        if self._use_primitives:
+            warnings.warn(
+                "`use_primitives` is set as True but the circuit will be ran without using "
+                "primitives. This is because an observable e.g. `qml.expval(qml.Hadamard(0))`"
+                " does not have a `pauli_rep`."
+            )
```
[the remark this answers is not in the document]
Comment (austingmhuang) [reply pennylane_qiskit/qiskit_device2.py:524]: That is a good point. The context is that I wanted some kind of warning that lets users know that running something like `qml.expval(qml.Z(0)), qml.expval(qml.Hadamard(0))` is potentially going to give an answer that is not well-supported/incorrect. I forgot about the fact that there are a lot of situations where you can arrive at not using primitives despite specifying that `use_primitives=True`.  It's probably better placed at the `split_measurement_types` transform, as you suggested.

File: pennylane_qiskit/qiskit_device2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane_qiskit/qiskit_device2.py:500]: Makes sense as a concern - I would suggest fixing the docstring and name of the function to reflect its modified behaviour, instead of avoiding using the function. Maybe it could be renamed as `split_execution_types` or `split_primitive_types`? The docstring will probably also need an update.

File: pennylane_qiskit/qiskit_device2.py [hunk 3]
```diff
@@ -556,6 +565,11 @@ def _execute_runtime_service(self, circuits, session):
 
         results = []
 
+        ### ToDo: ensure the measurements in circuit.measurements commute
+        ### Need to wait to implement transforms such as split non-commute, ham expand, etc.
+        ### [SC-62047]
+        ### Can't raise a warning/error because if we could figure out that the measurements
+        ### did not commute then we could just split it instead of raising a warning/error.
```
[the remark this answers is not in the document]
Comment (austingmhuang) [reply pennylane_qiskit/qiskit_device2.py:572]: I put this comment here instead of in the pre-processing step to emphasize that bugs that occur in `_execute_runtime_service` (specifically there) are a direct result of the measurements not commuting/other issues. 

For example if a user decides to set `use_primitives=False` and have a circuit that measures `qml.expval(qml.Z(0)), qml.expval(qml.X(0)), qml.expval(qml.Y(0))`, this will run without any error or warning but produce an inaccurate answer.

The point of the comment is to say that I would like to raise an error or warning in the event that the measurement is going to produce inaccurate results i.e. not diagonalized properly. But to note that if I could check if the diagonalization was incorrect then I could also diagnose what about it was incorrect and split the measurement instead of raising a warning/error. 

Right now I'm thinking about raising a warning that is a bit more general. For example, if there are measurement processes acting on the same wire, could raise a warning that says something like "Make sure these commute or else they might be inaccurate!!"

This warning falls out of the PR's scope, however, so I did not implement it but left it as a comment.

[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane_qiskit/qiskit_device2.py:572]: This feature branch won't be merged into `master` until that support is added, and users shouldn't be using feature branches that are being actively developed, so there's no need for any user-facing accommodations like warnings for that. 

>But to note that if I could check if the diagonalization was incorrect then I could also diagnose what about it was incorrect and split the measurement instead of raising a warning/error.

You shouldn't split the measurements here, because the post-processing knows what shape to expect based on what comes out of `preprocessing`. Anything that would split things into different tapes needs to occur in the `preprocessing` function so that the shapes of the output tuples from `execute` are as expected.

This function shouldn't worry about things like that at all - it should assume that any inputs it gets are valid, because pre-processing and validation is completed before this part of the execution pipeline.

[the remark this answers is not in the document]
Comment (austingmhuang) [reply pennylane_qiskit/qiskit_device2.py:572]: Ah I edited my comment cause I realized that was a bit weird to say but yeah, I wasn't intending to split the measurements there. I think I can clarify the comment a bit to mean something more like "Right now this function assumes that the inputs are valid, but since there are some transforms missing this isn't the case." Sort of a note to self type of comment.

File: tests/test_base_device.py [hunk 4]
```diff
@@ -1270,3 +1283,68 @@ def circuit():
...
+    @pytest.mark.parametrize(
+        "observable",
+        [
+            [qml.Hadamard(0), qml.PauliX(1)],
+            [qml.PauliZ(0), qml.Hadamard(1)],
+            [qml.PauliZ(0), qml.Hadamard(0)],
+        ],
+    )
+    @pytest.mark.filterwarnings("ignore::UserWarning")
+    def test_unsupported_observable_gives_accurate_answer(self, mocker, observable):
```
Comment (lillian542) [inline tests/test_base_device.py:1296]:
```suggestion
    def test_no_pauli_observable_gives_accurate_answer(self, mocker, observable):
```

**Review by lillian542 — APPROVED** [review]
Looks good to me! 🚀

File: pennylane_qiskit/qiskit_device2.py [hunk 5]
```diff
@@ -124,10 +124,12 @@ def accepted_sample_measurement(m: qml.measurements.MeasurementProcess) -> bool:
 
 
 @transform
-def split_measurement_types(
+def split_execution_types(
     tape: qml.tape.QuantumTape,
 ) -> (Sequence[qml.tape.QuantumTape], Callable):
-    """Split into separate tapes based on measurement type. Counts will use the
+    """Split into separate tapes based on measurement type. However, for `expval` and `var`
```
Comment (obliviateandsurrender) [inline pennylane_qiskit/qiskit_device2.py:130]:
```suggestion
    """Split into separate tapes based on measurement type. However, for ``expval`` and ``var``
```

File: pennylane_qiskit/qiskit_device2.py [hunk 6]
```diff
@@ -124,10 +124,12 @@ def accepted_sample_measurement(m: qml.measurements.MeasurementProcess) -> bool:
...
+    measurements, if the measured observable does not have a `pauli_rep`, it is split as a
```
Comment (obliviateandsurrender) [inline pennylane_qiskit/qiskit_device2.py:131]:
```suggestion
    measurements, if the measured observable does not have a ``pauli_rep``, it is split as a
```

File: pennylane_qiskit/qiskit_device2.py [hunk 7]
```diff
@@ -495,7 +505,10 @@ def execute_circuits(session):
                             f"Setting shot vector {circ.shots.shot_vector} is not supported for {self.name}."
                             f"The circuit will be run once with {circ.shots.total_shots} shots instead."
                         )
-                    if isinstance(circ.measurements[0], (ExpectationMP, VarianceMP)):
+                    if (
+                        isinstance(circ.measurements[0], (ExpectationMP, VarianceMP))
+                        and circ.measurements[0].obs.pauli_rep
```
Comment (obliviateandsurrender) [inline pennylane_qiskit/qiskit_device2.py:510]:
```suggestion
                        and getattr(circ.measurements[0], "pauli_rep", None)
```
I prefer `getattr` in such scenarios so that we don't get attribute not present errors in some edge cases.

File: tests/test_base_device.py [hunk 8]
```diff
@@ -1270,3 +1283,68 @@ def circuit():
...
+
+    @pytest.mark.parametrize(
+        "observable",
+        [
+            [qml.Hadamard(0), qml.PauliX(1)],
+            [qml.PauliZ(0), qml.Hadamard(1)],
+            [qml.PauliZ(0), qml.Hadamard(0)],
+        ],
+    )
+    @pytest.mark.filterwarnings("ignore::UserWarning")
```
Comment (obliviateandsurrender) [inline tests/test_base_device.py:1295]: Why do we have this? 🤔

**Review by obliviateandsurrender — APPROVED** [review]
This should be good to go once all tests pass!
